bhttp/src/http_compat/decode.rs

```diff
@@ -0,0 +1,501 @@
+///
+/// This decoder works with any type that implements [futures::AsyncRead],
+/// making it suitable for various input sources like network streams or files.
+///
+/// # Important Notes
+///
+/// - Due to limitations in the http crate, any Informational Response (1xx) status
+///   codes in HTTP responses will be ignored during encoding.
+/// - In our support for Indeterminate-Length Binary Encoding, the body is treated
+///   as a continuous binary stream, which means HTTP frames might be reorganized
+///   or split. Applications should implement their own framing methods and should
+///   not assume that Frame encodings from the http crate will remain the same
+///   after encoding and decoding.
+pub struct BhttpDecoder<R> {
+    reader: R,
+}
+
+impl<R> BhttpDecoder<R> {
+    /// Create a new BHTTP decoder
+    ///
+    /// # Arguments
+    ///
+    /// * `reader` - The async reader to decode BHTTP data from
+    pub fn new(reader: R) -> Self {
+        Self { reader }
+    }
+}
+
+impl<R> BhttpDecoder<R>
+where
+    R: FuturesAsyncRead + Unpin + Send + 'static,
+{
+    /// Decode an HTTP message from BHTTP format, returning either a request or response
+    ///
+    /// This method performs the complete decoding process:
+    /// 1. Reads the message header
+    /// 2. Determines if it's a request or response
+    /// 3. Constructs the appropriate HTTP type with a streaming body
+    ///
+    /// # Returns
+    ///
+    /// A result containing either an HTTP request or response with a streaming body,
+    /// or an error if decoding fails.
+    pub async fn decode_message(self) -> Res<HttpRequestOrResponse<R>> {
+        let mut async_message = Message::async_read(self.reader);
+
+        let header = async_message.header().await?;
+        let parts = HttpRequestOrResponseParts::try_from(header)?;
+
+        match parts {
+            HttpRequestOrResponseParts::Request(parts) => Ok(HttpRequestOrResponse::Request(
+                Request::from_parts(parts, BhttpBody::new(async_message)),
+            )),
+            HttpRequestOrResponseParts::Response(parts) => Ok(HttpRequestOrResponse::Response(
+                Response::from_parts(parts, BhttpBody::new(async_message)),
+            )),
+        }
+    }
+}
+
+impl TryFrom<crate::Header> for HttpRequestOrResponseParts {
+    type Error = Error;
+
+    /// Convert a BHTTP header into either HTTP request or response parts
+    ///
+    /// This method parses the control data and field section from a BHTTP header
+    /// and constructs the corresponding HTTP request or response parts.
+    ///
+    /// # Arguments
+    ///
+    /// * `header` - The BHTTP header to convert
+    ///
+    /// # Returns
+    ///
+    /// Either HTTP request parts or response parts, depending on the control data,
+    /// or an error if parsing fails.
+    fn try_from(header: crate::Header) -> Res<Self> {
+        let control_data = header.control();
+
+        if control_data.is_request() {
```

Comment:
This function needs a bit of factoring into parts.  It's pretty long and involved.